Newlines in return type cause scalafmt to crash with `error: Unable to format file due to bug in scalafmt`

## Configuration (required) ##
```
version = 3.4.3
runner.dialect = scala213

newlines {
  neverInResultType = true
}

```
## Command-line parameters (required) ##

When I run scalafmt via CLI like this: `scalafmt src/Inliner.scala`

## Steps

Given code like this:
`src/Inliner.scala`
```scala
object Inliner {

  def zipMethodsWithTheirInlinedVersionsFromParent(
      prog: EOProg[EOExprOnly]
  ): EitherNel[
    String,
    Map[
      EONamedBnd,
      ObjectTree[
        (AnalysisInfo, AnalysisInfo)
      ]
    ]
  ] = {
    ???
  }

}

```

## Problem

`scalafmt` crashes with an error message:
```
org.scalafmt.cli.FailedToFormat: /home/nikololiahim/scala/scalafmt-bug/src/Inliner.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: /home/nikololiahim/scala/scalafmt-bug/src/Inliner.scala:5: error: Unable to format file due to bug in scalafmt
  ): EitherNel[
              ^
        at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$7(Scalafmt.scala:161)
        at scala.util.Success.flatMap(Try.scala:258)
        at org.scalafmt.Scalafmt$.$anonfun$doFormatOne$4(Scalafmt.scala:156)
        at scala.meta.parsers.Parsed.fold(Errors.scala:12)
        at scala.meta.parsers.Parsed.fold$(Errors.scala:11)
        at scala.meta.parsers.Parsed$Success.fold(Errors.scala:26)
        at org.scalafmt.Scalafmt$.org$scalafmt$Scalafmt$$doFormatOne(Scalafmt.scala:152)
        at org.scalafmt.Scalafmt$.doFormat(Scalafmt.scala:126)
        at org.scalafmt.Scalafmt$.formatCodeWithStyle(Scalafmt.scala:91)
        at org.scalafmt.Scalafmt$.$anonfun$formatCode$3(Scalafmt.scala:78)
        at scala.util.Success.fold(Try.scala:281)
        at org.scalafmt.Scalafmt$.formatCode(Scalafmt.scala:78)
        at org.scalafmt.cli.ScalafmtCoreRunner$.unsafeHandleFile(ScalafmtCoreRunner.scala:69)
        at org.scalafmt.cli.ScalafmtCoreRunner$.handleFile(ScalafmtCoreRunner.scala:53)
        at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$5(ScalafmtCoreRunner.scala:36)
        at org.scalafmt.cli.ScalafmtCoreRunner$.$anonfun$run$5$adapted(ScalafmtCoreRunner.scala:35)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1293)
        at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:938)
        at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:52)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:97)
        at scala.collection.parallel.Task.tryLeaf(Tasks.scala:55)
        at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:49)
        at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:935)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute(Tasks.scala:152)
        at scala.collection.parallel.AdaptiveWorkStealingTasks$AWSTWrappedTask.compute$(Tasks.scala:148)
        at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$AWSFJTWrappedTask.compute(Tasks.scala:304)
        at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

## Expectation
I would expect this file to be formatted somehow.

## Notes
Setting `newlines.neverInResultType` to `false` causes the error to go away. The file remains unchanged, though.
Same happens when I use `sbt-scalafmt` v2.4.6.

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -70,6 +70,7 @@ class Router(formatOps: FormatOps) {
     prev,
     next,
     tokenBefore,
+    getLastNonTrivial,
     prevNonComment,
     nextNonComment,
     prevNonCommentSameLine,
@@ -1398,7 +1399,7 @@ class Router(formatOps: FormatOps) {
       case FormatToken(left, _: T.Colon, DefDefReturnTypeRight(returnType))
           if style.newlines.sometimesBeforeColonInMethodReturnType
             || left.is[T.Comment] && newlines != 0 =>
-        val expireFt = tokens.getLastNonTrivial(returnType)
+        val expireFt = getLastNonTrivial(returnType)
         val expire = expireFt.left
         val sameLineSplit = Space(endsWithSymbolIdent(left))
         val bopSplits = style.newlines.getBeforeOpenParenDefnSite.map { x =>
@@ -1431,12 +1432,9 @@ class Router(formatOps: FormatOps) {
         }
       case FormatToken(_: T.Colon, _, DefDefReturnTypeLeft(returnType))
           if style.newlines.avoidInResultType =>
-        val expire = returnType match {
-          case Type.Refine(_, headStat :: _) =>
-            tokens.tokenJustBefore(headStat).left
-          case t => getLastNonTrivialToken(t)
-        }
-        Seq(Split(Space, 0).withPolicy(SingleLineBlock(expire, okSLC = true)))
+        val expire = getLastNonTrivialToken(returnType)
+        val policy = PenalizeAllNewlines(expire, Constants.ShouldBeNewline)
+        Seq(Split(Space, 0).withPolicy(policy).withOptimalToken(expire))
 
       case FormatToken(T.LeftParen(), T.LeftBrace(), _) =>
         Seq(
@@ -1531,18 +1529,20 @@ class Router(formatOps: FormatOps) {
             val lastFT = tokens.getLast(nextArg)
             val loEnd = leftOwner.tokens.last.end
             val oneline = binPack == BinPack.Unsafe.Oneline
-            val nextCommaOrParenOneline = if (oneline) {
+            val nextCommaOrParen =
               findFirst(lastFT, loEnd) {
                 case FormatToken(_, _: T.Comma, _) => true
                 case FormatToken(_, RightParenOrBracket(), _) => true
                 case _ => false
               }
-            } else None
-            val optFT = nextCommaOrParenOneline match {
-              case Some(ft @ FormatToken(_, _: T.Comma, _)) => ft
+            val optFT = nextCommaOrParen match {
+              case Some(ft @ FormatToken(_, _: T.Comma, _)) if oneline => ft
+              case Some(FormatToken(_, _: T.RightParen, _))
+                  if style.newlines.avoidInResultType =>
+                defDefReturnType(leftOwner).fold(lastFT)(getLastNonTrivial)
               case _ => lastFT
             }
-            val nlPolicy = nextCommaOrParenOneline match {
+            val nlPolicy = (if (oneline) nextCommaOrParen else None) match {
               case Some(FormatToken(_, t: T.Comma, _)) =>
                 if (callSite) splitOneArgPerLineAfterCommaOnBreak(t)
                 else delayedBreakPolicyFor(t)(decideNewlinesOnlyAfterClose)

--- a/scalafmt-tests/src/test/resources/scalajs/DefDef.stat
+++ b/scalafmt-tests/src/test/resources/scalajs/DefDef.stat
@@ -38,7 +38,72 @@ def a =
  js.native
 >>>
 def a = js.native
-<<< default penalty #247
+<<< default penalty #247, binPack=always, dangle
+binPack.unsafeDefnSite = always
+danglingParentheses.defnSite = true
+===
+{
+  def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
+}
+>>>
+{
+  def fetch(info: RequestInfo,
+      init: RequestInit = null): js.Promise[Response] = js.native
+}
+<<< default penalty #247, binPack=always, !dangle
+binPack.unsafeDefnSite = always
+danglingParentheses.defnSite = false
+===
+{
+  def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
+}
+>>>
+{
+  def fetch(info: RequestInfo,
+      init: RequestInit = null): js.Promise[Response] = js.native
+}
+<<< default penalty #247, binPack=always, !dangle, avoidInResultType
+binPack.unsafeDefnSite = always
+newlines.avoidInResultType = true
+danglingParentheses.defnSite = false
+===
+{
+  def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
+}
+>>>
+{
+  def fetch(info: RequestInfo,
+      init: RequestInit = null): js.Promise[Response] = js.native
+}
+<<< default penalty #247, binPack=oneline, dangle
+binPack.unsafeDefnSite = oneline
+danglingParentheses.defnSite = true
+===
+{
+  def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
+}
+>>>
+{
+  def fetch(info: RequestInfo,
+      init: RequestInit = null): js.Promise[Response] = js.native
+}
+<<< default penalty #247, binPack=oneline, !dangle
+binPack.unsafeDefnSite = oneline
+danglingParentheses.defnSite = false
+===
+{
+  def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
+}
+>>>
+{
+  def fetch(info: RequestInfo,
+      init: RequestInit = null): js.Promise[Response] = js.native
+}
+<<< default penalty #247, binPack=oneline, !dangle, avoidInResultType
+binPack.unsafeDefnSite = oneline
+newlines.avoidInResultType = true
+danglingParentheses.defnSite = false
+===
 {
   def fetch(info: RequestInfo, init: RequestInit = null): js.Promise[Response] = js.native
 }

--- a/scalafmt-tests/src/test/resources/unit/DefDef.stat
+++ b/scalafmt-tests/src/test/resources/unit/DefDef.stat
@@ -207,9 +207,21 @@ object a {
   }
 }
 >>>
-BestFirstSearch:323 Failed to format
-UNABLE TO FORMAT,
-tok=[∙String[109:120]
+object a {
+  def zipMethodsWithTheirInlinedVersionsFromParent(
+      prog: EOProg[EOExprOnly]
+  ): EitherNel[
+      String,
+      Map[
+          EONamedBnd,
+          ObjectTree[
+              (AnalysisInfo, AnalysisInfo)
+          ]
+      ]
+  ] = {
+    ???
+  }
+}
 <<< #3156 avoidInResultType=true, !cfgstyle
 newlines.avoidInResultType = true
 optIn.configStyleArguments = false

--- a/scalafmt-tests/src/test/resources/unit/DefDef.stat
+++ b/scalafmt-tests/src/test/resources/unit/DefDef.stat
@@ -187,3 +187,53 @@ object a {
         type Out = RouteTestResult
       } = TildeArrow.injectIntoRoute
 }
+<<< #3156 avoidInResultType=true, cfgstyle
+newlines.avoidInResultType = true
+optIn.configStyleArguments = true
+===
+object a {
+  def zipMethodsWithTheirInlinedVersionsFromParent(
+      prog: EOProg[EOExprOnly]
+  ): EitherNel[
+    String,
+    Map[
+      EONamedBnd,
+      ObjectTree[
+        (AnalysisInfo, AnalysisInfo)
+      ]
+    ]
+  ] = {
+    ???
+  }
+}
+>>>
+BestFirstSearch:323 Failed to format
+UNABLE TO FORMAT,
+tok=[∙String[109:120]
+<<< #3156 avoidInResultType=true, !cfgstyle
+newlines.avoidInResultType = true
+optIn.configStyleArguments = false
+===
+object a {
+  def zipMethodsWithTheirInlinedVersionsFromParent(
+      prog: EOProg[EOExprOnly]
+  ): EitherNel[
+    String,
+    Map[
+      EONamedBnd,
+      ObjectTree[
+        (AnalysisInfo, AnalysisInfo)
+      ]
+    ]
+  ] = {
+    ???
+  }
+}
+>>>
+object a {
+  def zipMethodsWithTheirInlinedVersionsFromParent(
+      prog: EOProg[EOExprOnly])
+      : EitherNel[String, Map[EONamedBnd, ObjectTree[(AnalysisInfo, AnalysisInfo)]]] = {
+    ???
+  }
+}

